File: Algorithms/LinkedList-Algorithm/Floyd_Cylce_detection/floyd_cycle_detection.kt
class LinkedList<T> {
    class Node<T>(val value: T, var next: Node<T>? = null)

    var head: Node<T>? = null
    private lateinit var tail: Node<T>

    fun insert(item: T) {
        if (head == null) {
            tail = Node(item)
            head = tail
        } else {
            tail.next = Node(item)
            tail = tail.next!!
        }
    }

    fun detectLoop(): Boolean {
        var slowPointer = head?.next
        var fastPointer = head?.next?.next

        while (slowPointer != null && fastPointer != null) {
            if (slowPointer == fastPointer) return true
            slowPointer = slowPointer.next
            fastPointer = fastPointer.next?.next
        }

        return false
    }

    fun forceLoop() {
        tail.next = head
    }
}

fun main() {
    val linkedList = LinkedList<Int>()
    linkedList.apply {
        insert(1)
    }
    println("linkedList has loop: " + linkedList.detectLoop())

    linkedList.forceLoop()
    println("linkedList has loop: " + linkedList.detectLoop())
}

File: Algorithms/LinkedList-Algorithm/Floyd_Cylce_detection/floyd_cycle_detection.py
 class Node:  
	def __init__(self, data): 
		self.data = data 
		self.next = None

class LinkedList:  
	def __init__(self): 
		self.head = None


	def push(self, new_data): 
		new_node = Node(new_data) 
		new_node.next = self.head 
		self.head = new_node 


	def printList(self): 
		temp = self.head 
		while(temp): 
			print (temp.data, end =" ") 
			temp = temp.next


	def detectLoop(self): 
		s = set() 
		temp = self.head 
		while (temp): 
			if (temp in s): 
				return True

			s.add(temp) 
			temp = temp.next
		
	
		return False



# Driver program for testing 
llist = LinkedList() 
llist.push(10) 
llist.push(4) 
llist.push(15) 
llist.push(40) 

# Create a loop for testing 
llist.head.next.next.next.next = llist.head; 

if( llist.detectLoop()): 
	print ("Loop found") 
else : 
	print ("No Loop ") 

